Dark nav for standalone Gnav

nishantka opened the pull request:
Dark Mode implementation for Standalone globalnav and footer. It should not impact milo consumers as dark mode is enabled using site config (a new property `theme` which should be `dark`).

Resolves: [MWPW-155090](https://jira.corp.adobe.com/browse/MWPW-155090)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/?martech=off
- After: https://dark-nav--milo--nishantka.hlx.page/?martech=off

QA: 
- Before: https://adobecom.github.io/nav-consumer/navigation.html?env=stage&navbranch=stage
- After: https://adobecom.github.io/nav-consumer/navigation.html?env=stage&navbranch=dark-nav&theme=dark

overmyheadandbody reviewed (commented):
It looks like `dark-nav.css` has color updates based on CSS custom properties. Instead of redefining selectors, why not redefine variables? Something like:

- you keep things mostly as is, you just rename variables to remove the `--light` modifier and replace throughout the style definitions;
- either you create a new file (and import it once you figure out dark mode is needed) or add the dark styles where global navigation variables are already defined (but be sure to investigate the performance hit of adding additional bytes for all use-cases). Here you can scope the variables inside your modifier class;
- you'd then get OOTB color updates based on your modifier class without having to duplicate any selectors, making things more maintainable in the long run.

As for the modifier class, all global navigation/footer styles are prefixed with `.feds`, why not call it `.feds--dark` or something similar?

nishantka commented:
> It looks like `dark-nav.css` has color updates based on CSS custom properties. Instead of redefining selectors, why not redefine variables? Something like:
> 
> * you keep things mostly as is, you just rename variables to remove the `--light` modifier and replace throughout the style definitions;
> * either you create a new file (and import it once you figure out dark mode is needed) or add the dark styles where global navigation variables are already defined (but be sure to investigate the performance hit of adding additional bytes for all use-cases). Here you can scope the variables inside your modifier class;
> * you'd then get OOTB color updates based on your modifier class without having to duplicate any selectors, making things more maintainable in the long run.
> 
> As for the modifier class, all global navigation/footer styles are prefixed with `.feds`, why not call it `.feds--dark` or something similar?

Thanks @overmyheadandbody,
I can work on updating all the variables/usages with `--light` and make them generic to avoid duplicate selectors.
There are some new selectors required only for dark mode. We can keep them in `dark-nav.css` along with dark mode variables and import as required.
Also, there are some places where we can define new variables in place of colors.
My only concern is the risk it poses as it vastly impacts Milo GNav for dark mode theme required (as of now) only for Standalone GNav.

sharmrj commented on libs/blocks/global-navigation/dark-nav.css:
```diff
@@ -0,0 +1,492 @@
+}
+
+.darknav .feds-breadcrumbs [aria-current] {
+  color: #DBDBDB;
```
There seem to be plenty of colors repeated throughout the file that already have variables associated with them. For example here `#DBDBDB` is `--feds-color-link--dark`/`--feds-color-headline--dark`. You may or may not want to replace these repetitions with variables depending on whether changing one color should change the other instances of the same color.

narcis-radu reviewed (dismissed):
I noticed some issues and need to request some changes:
1. color contrast issues - expand Creativity & Design and check the "Register now" in the promo. The link is weird looking in light mode as well, should be fixed.
2. at times the navigation would get rendered (lightmode) and then the darkmode gets applied, resulting in a weird color transition. I am not able to reproduce this consistently though, however I was able to get a screenshot.
<img width="1358" alt="Screenshot 2024-08-14 at 16 14 25" src="https://github.com/user-attachments/assets/6dd2e26f-9350-42f5-ba39-399155645f62">

nishantka commented:
> I noticed some issues and need to request some changes:
> 
> 1. color contrast issues - expand Creativity & Design and check the "Register now" in the promo. The link is weird looking in light mode as well, should be fixed.
> 2. at times the navigation would get rendered (lightmode) and then the darkmode gets applied, resulting in a weird color transition. I am not able to reproduce this consistently though, however I was able to get a screenshot.
> 
> <img alt="Screenshot 2024-08-14 at 16 14 25" width="1358" src="https://private-user-images.githubusercontent.com/408175/357827405-6dd2e26f-9350-42f5-ba39-399155645f62.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MDIyNjksIm5iZiI6MTcyMzgwMTk2OSwicGF0aCI6Ii80MDgxNzUvMzU3ODI3NDA1LTZkZDJlMjZmLTkzNTAtNDJmNS1iYTM5LTM5OTE1NTY0NWY2Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxNlQwOTUyNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNDBlODI0MDU1YWQ2ZDZiYWFmZjEzY2JiZTU2MmI3ZDU2ZTQ1YWQ1MWVhYWEwNTdkYmZlMTk0NGM3OWNkNzEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-5Dp4lbZ_g3bJJ-98SS8cJdQz6HoMTl40_y3S1zub8M">

Hi @narcis-radu, thanks for review.
1. color contrast issues - I've updated the promo link color for dark theme
2. I've added `header.global-navigation.feds--dark` in `styles.css` to hide navigation until `dark-nav.css` is loaded -which makes it visible again. It should be fixed now.

nishantka commented:
> It looks like `dark-nav.css` has color updates based on CSS custom properties. Instead of redefining selectors, why not redefine variables? Something like:
> 
> * you keep things mostly as is, you just rename variables to remove the `--light` modifier and replace throughout the style definitions;
> * either you create a new file (and import it once you figure out dark mode is needed) or add the dark styles where global navigation variables are already defined (but be sure to investigate the performance hit of adding additional bytes for all use-cases). Here you can scope the variables inside your modifier class;
> * you'd then get OOTB color updates based on your modifier class without having to duplicate any selectors, making things more maintainable in the long run.
> 
> As for the modifier class, all global navigation/footer styles are prefixed with `.feds`, why not call it `.feds--dark` or something similar?

Hi @overmyheadandbody,
I've removed variables using `--light` from GNav and added override variables in `dark-nav.css` to remove most of the duplicate selectors. Some selectors are kept to avoid adding them unnecessarily in milo at this point. Also, some selectors are kept as few file specifc variables which loads later will override dark theme variables.  Now, we will be loading `dark-nav.css` once `base.css` is loaded
Also, I've updated the classname from `darknav` to `feds--dark`.

prativas22 commented:
https://adobecom.github.io/nav-consumer/navigation.html?env=stage&navbranch=dark-nav&theme=dark 
![image](https://github.com/user-attachments/assets/ab9156d1-513a-4684-98b2-a726e5948dc6)
![image](https://github.com/user-attachments/assets/4c9d7cbd-aba9-4a14-ae2c-433727a20b91)
![image](https://github.com/user-attachments/assets/b9c761ca-cdae-43f1-88ca-3be78d4007a4)
![image](https://github.com/user-attachments/assets/39b7b177-f5cd-4eff-8fb7-156d22b697e3)

https://adobecom.github.io/nav-consumer/navigation.html?env=stage&navbranch=dark-nav&theme=dark&authoringpath=/federal/dev/gnavtest/dcgnav/gnav/localnavacrobatteams
![image](https://github.com/user-attachments/assets/d28f958a-2ca6-4206-bd13-9e04b6e7803f)
Localized page:
![image](https://github.com/user-attachments/assets/3468364f-6455-472c-a1b6-e8c0781a77c7)
Sign in session :
![image](https://github.com/user-attachments/assets/8727d562-ee45-48dc-95ec-00ac8fd402de)

GDPR check:
https://adobecom.github.io/nav-consumer/navigation.html?env=stage&navbranch=dark-nav&theme=dark&authoringpath=/federal/dev/gnavtest/ccgnav/designshortform
![image](https://github.com/user-attachments/assets/5db77050-4faf-455b-ad12-eda760be7847)

overmyheadandbody commented on libs/blocks/global-navigation/utilities/utilities.js:
```diff
@@ -156,12 +161,22 @@ export function loadStyles(url) {
 export async function loadBaseStyles() {
-  const url = rootPath('base.css');
-  await loadStyles(url);
+  if (isDarkMode()) {
+    new Promise((resolve) => { loadStyle(rootPath('base.css'), resolve); })
+      .then(() => loadStyles(rootPath('dark-nav.css')));
+  } else {
+    const url = rootPath('base.css');
+    await loadStyles(url);
+  }
```
Maybe you can define an array of Promises. Initial entry it just base. Conditionally, you can then add the dark-nav one. Then run Promise.all. In theory, this would parallelize the calls and apply the theme a bit quicker, lowering the risk of FOUC; or rather flash of unthemed content in this case 🙂 And could also speed things along a bit. This is the current timeline of these files:
<img width="530" alt="Screenshot 2024-08-20 at 17 07 31" src="https://github.com/user-attachments/assets/1a4cf910-941d-477e-8f6e-fd1baccddede">

nishantka replied:
Since, now we're making nav `visible` in `dark-nav.css` it is made to load after `base.css`. If we load them in parallel then if `dark-nav.css` completes loading before `base.css` it will have FOUC as seen below:

<img width="1792" alt="Screenshot 2024-08-21 at 1 26 17 PM" src="https://github.com/user-attachments/assets/f912fe53-28bf-4b7e-8fc4-adda2243db9a">

overmyheadandbody commented on libs/navigation/navigation.css:
```diff
@@ -24,7 +24,16 @@
 }
  
-header.global-navigation {
+header.global-navigation, header.global-navigation.feds--dark {
```
Is the repetition really needed, since, both match `header.global-navigation`?

overmyheadandbody reviewed (commented):
Thanks for integrating all the feedback, this looks much better and is going to be a lot more scalable. No big objections, just a few notes.

nishantka commented on libs/styles/styles.css:
```diff
@@ -703,7 +703,7 @@ header:not(.global-navigation) ~ main {
   all of the ':not(.global-navigation)' rules can be entirely removed.
  */
-header.global-navigation {
+header.global-navigation, header.global-navigation.feds--dark {
```
This is added to fix the issue which @narcis-radu reported above.
`header.global-navigation` is made visible in `base.css` - In dark made this sometime makes the Nav visible with light mode momentarily until `dark-nav.css` completes loading.
So, this was added to keep nav hidden for dark mode with a more specific selector.

nishantka commented on libs/blocks/global-navigation/dark-nav.css:
```diff
@@ -0,0 +1,262 @@
+.feds-cta--primary {
+  background-color: #066ce7;
+  border-color: #066ce7;
+  color: rgb(255, 255, 255);
```
I've updated this to make it more consistent.

nishantka commented on libs/blocks/global-navigation/dark-nav.css:
```diff
@@ -0,0 +1,262 @@
+}
+
+.feds--dark .feds-cta--secondary {
+  background-color: #111;
```
Modified to reuse the variables here

spadmasa commented:
Validations are done on the pr details are updated in subtask in the ticket
https://jira.corp.adobe.com/browse/MWPW-156618
